src/libraries/Microsoft.PowerFx.Core/Public/Config/SymbolTable.cs

```diff
@@ -67,27 +40,52 @@ private void ValidateName(string name)
             {
                 throw new ArgumentException("Invalid name: ${name}");
             }
-        }
-
+        }
+
+        private int _slotCounter = 0;
+
         /// <summary>
         /// Provide variable for binding only.
         /// Value must be provided at runtime.
         /// </summary>
         /// <param name="name"></param>
-        /// <param name="type"></param>
-        public void AddVariable(string name, FormulaType type)
-        {
+        /// <param name="type"></param>
+        /// <param name="mutable"></param>
+        /// <param name="displayName"></param>
+        public ISymbolSlot AddVariable(string name, FormulaType type, bool mutable = false, string displayName = null)
+        {
+            if (displayName != null)
+            {
+                // Include parameter so that it's not a breaking change when we enable.
+                // https://github.com/microsoft/Power-Fx/issues/779
+                throw new NotImplementedException("DisplayName support for variables not implemented yet");
+            }
+
             Inc();
-            ValidateName(name);
+            ValidateName(name);
+
+            if (_variables.ContainsKey(name))
+            {
+                throw new InvalidOperationException($"{name} is already defined");
+            }
+
+            var data = new NameSymbol(name, mutable)
+            {
+                Parent = this,
+                SlotIndex = _slotCounter
+            };
+            _slotCounter++; // $$$ Compact and make dense. 

```

Comment:
> $$$ Compact and make dense. [](http://example.com/codeflow?start=30&length=28)

Not sure of what you really mean here

Reply:
The $$$ are todo markers that I still need to resolve.